File: app/src/main/java/net/opendasharchive/openarchive/fragments/MediaListFragment.java
package net.opendasharchive.openarchive.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import net.opendasharchive.openarchive.R;
import net.opendasharchive.openarchive.db.Media;
import net.opendasharchive.openarchive.db.MediaAdapter;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class MediaListFragment extends Fragment {


    protected LinearLayout mMediaContainer;
    MediaAdapter mMediaAdapter;
    protected static final String TAG = "RecyclerViewFragment";

    protected long mProjectId = -1;
    protected long mStatus = Media.STATUS_UPLOADING;
    protected long[] mStatuses = {Media.STATUS_UPLOADING,Media.STATUS_QUEUED};


    /**
     * Mandatory empty constructor for the fragment manager to instantiate the
     * fragment (e.g. upon screen orientation changes).
     */
    public MediaListFragment() {
    }

    public List<Media> getMediaList ()
    {
        return mMediaAdapter.getMediaList();
    }

    public void setStatus (long status) {
        mStatus = status;
    }

    public void setProjectId (long projectId)
    {
        mProjectId = projectId;
    }

    public void updateItem (long mediaId, long progress)
    {
        if (mMediaAdapter != null)
            mMediaAdapter.updateItem(mediaId, progress);

    }

    public void stopBatchMode ()
    {
        if (mMediaAdapter != null && mMediaAdapter.getActionMode() != null)
            mMediaAdapter.getActionMode().finish();
    }

    public void refresh ()
    {
        if (mMediaAdapter != null)
        {
            List<Media> listMedia = null;

            if (mProjectId == -1)
            {
                listMedia = Media.getMediaByStatus(mStatuses);

            }
            else
            {
                listMedia = Media.getMediaByProject(mProjectId);
            }

            mMediaAdapter.updateData(listMedia);

        }

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_media_list, container, false);
        rootView.setTag(TAG);

        mMediaContainer = rootView.findViewById(R.id.mediacontainer);

        //mRecyclerView = rootView.findViewById(R.id.recyclerview);
        RecyclerView rView = new RecyclerView(getContext());
        rView.setLayoutManager(new LinearLayoutManager(getActivity()));
        rView.setHasFixedSize(true);

        mMediaContainer.addView(rView);

        List<Media> listMedia = null;

        if (mProjectId == -1)
        {
            listMedia = Media.getMediaByStatus(mStatuses);

        }
        else
        {
            listMedia = Media.getMediaByProject(mProjectId);

            for (Media media : listMedia)
            {
                if (media.status == Media.STATUS_LOCAL)
                {

                    break;
                }
            }
        }

        mMediaAdapter = new MediaAdapter(getActivity(), R.layout.activity_media_list_row_short, listMedia, rView, new OnStartDragListener() {
            @Override
            public void onStartDrag(RecyclerView.ViewHolder viewHolder) {

                mItemTouchHelper.startDrag(viewHolder);

            }
        });
        mMediaAdapter.setDoImageFade(false);
        rView.setAdapter(mMediaAdapter);
        mItemTouchHelper.attachToRecyclerView(rView);

        return rootView;
    }

    private ItemTouchHelper mItemTouchHelper = new ItemTouchHelper(new ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP|ItemTouchHelper.DOWN,ItemTouchHelper.END|ItemTouchHelper.START) {
        @Override
        public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {

            mMediaAdapter.onItemMove(viewHolder.getAdapterPosition(),
                    target.getAdapterPosition());

            return true;
        }

        @Override
        public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {

            mMediaAdapter.onItemDismiss(viewHolder.getAdapterPosition());
        }
    });

    public void setEditMode (boolean isEditMode)
    {
        mMediaAdapter.setEditMode(isEditMode);
    }

    public interface OnStartDragListener {

        /**
         * Called when a view is requesting a start of a drag.
         *
         * @param viewHolder The holder of the view to drag.
         */
        void onStartDrag(RecyclerView.ViewHolder viewHolder);
    }

}

File: app/src/main/java/net/opendasharchive/openarchive/features/media/list/MediaListFragment.kt
package net.opendasharchive.openarchive.features.media.list

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import net.opendasharchive.openarchive.R
import net.opendasharchive.openarchive.databinding.FragmentMediaListBinding
import net.opendasharchive.openarchive.db.Media
import net.opendasharchive.openarchive.db.Media.Companion.getMediaByProject
import net.opendasharchive.openarchive.db.Media.Companion.getMediaByStatus
import net.opendasharchive.openarchive.db.MediaAdapter
import net.opendasharchive.openarchive.util.Constants.EMPTY_ID

open class MediaListFragment : Fragment() {

    private var mProjectId: Long = -1
    private var mStatus = Media.STATUS_UPLOADING.toLong()
    private var mStatuses = longArrayOf(
        Media.STATUS_UPLOADING.toLong(),
        Media.STATUS_QUEUED.toLong(), Media.STATUS_ERROR.toLong()
    )
    open var mMediaAdapter: MediaAdapter? = null

    private var _mBinding: FragmentMediaListBinding? = null
    private lateinit var viewModel: MediaListViewModel

    private val mItemTouchHelper = ItemTouchHelper(object : ItemTouchHelper.SimpleCallback(
        ItemTouchHelper.UP or ItemTouchHelper.DOWN,
        ItemTouchHelper.END or ItemTouchHelper.START
    ) {
        override fun onMove(
            recyclerView: RecyclerView,
            viewHolder: RecyclerView.ViewHolder,
            target: RecyclerView.ViewHolder
        ): Boolean {
            mMediaAdapter?.onItemMove(
                viewHolder.adapterPosition,
                target.adapterPosition
            )
            return true
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
            mMediaAdapter?.onItemDismiss(viewHolder.adapterPosition)
        }
    })

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _mBinding = FragmentMediaListBinding.inflate(inflater, container, false)
        _mBinding?.root?.tag = TAG
        viewModel = ViewModelProvider(this).get(MediaListViewModel::class.java)
        observeData()
        viewModel.getMediaList(mProjectId, mStatuses)
        return _mBinding?.root
    }

    private fun observeData() {
        viewModel.mediaList.observe(viewLifecycleOwner, Observer {
            if (mProjectId != EMPTY_ID) {
                it?.forEach { media ->
                    if (media.status == Media.STATUS_LOCAL) {
                        return@forEach
                    }
                }
                initLayout(it ?: listOf())
            }
        })
    }

    private fun initLayout(mediaList: List<Media>) {
        val rView = RecyclerView(requireContext())
        rView.layoutManager = LinearLayoutManager(activity)
        val itemDecorator = DividerItemDecoration(context, DividerItemDecoration.VERTICAL)
        itemDecorator.setDrawable(
            activity?.let {
                ContextCompat.getDrawable(
                    it,
                    R.drawable.divider
                )
            }!!
        )
        rView.addItemDecoration(itemDecorator)
        rView.setHasFixedSize(true)
        _mBinding?.mediacontainer?.addView(rView)

        val listMediaArray = ArrayList(mediaList)

        mMediaAdapter =
            MediaAdapter(
                requireContext(),
                R.layout.activity_media_list_row_short,
                listMediaArray,
                rView,
                object : OnStartDragListener {
                    override fun onStartDrag(viewHolder: RecyclerView.ViewHolder?) {
                        if (viewHolder != null) mItemTouchHelper.startDrag(viewHolder)
                    }

                }, onDelete = {})
        mMediaAdapter?.setDoImageFade(false)
        rView.adapter = mMediaAdapter
        mItemTouchHelper.attachToRecyclerView(rView)
    }

    fun getMediaList(): List<Media>? {
        return mMediaAdapter?.getMediaList()
    }

    open fun setStatus(status: Long) {
        mStatus = status
    }

    open fun getProjectId(): Long {
        return mProjectId
    }

    fun setProjectId(projectId: Long) {
        mProjectId = projectId
    }

    open fun updateItem(mediaId: Long, progress: Long) {
        mMediaAdapter?.updateItem(mediaId, progress)
    }

    fun stopBatchMode() {
        mMediaAdapter?.getActionMode()?.finish()
    }

    fun setEditMode(isEditMode: Boolean) {
        mMediaAdapter?.setEditMode(isEditMode)
    }

    open fun refresh() {
        val listMedia: List<Media>? = if (mProjectId == -1L) {
            getMediaByStatus(mStatuses, Media.ORDER_PRIORITY)
        } else {
            getMediaByProject(mProjectId)
        }

        val filterList = listMedia?.filter { it ->
            it.status == Media.STATUS_UPLOADING || it.status == Media.STATUS_QUEUED
        }

        filterList?.let {
            val listMediaArray = ArrayList(it)
            mMediaAdapter?.updateData(listMediaArray)
        } ?: run {
            mMediaAdapter?.updateData(arrayListOf())
        }
    }

    open fun getUploadingCounter(): Int {
        val listMedia: List<Media>? = if (mProjectId == -1L) {
            getMediaByStatus(mStatuses, Media.ORDER_PRIORITY)
        } else {
            getMediaByProject(mProjectId)
        }

        val filterList = listMedia?.filter { it ->
            it.status == Media.STATUS_QUEUED || it.status == Media.STATUS_UPLOADING
        }

        return filterList?.size ?: 0
    }

    override fun onResume() {
        super.onResume()
        refresh()
    }

    interface OnStartDragListener {
        /**
         * Called when a view is requesting a start of a drag.
         *
         * @param viewHolder The holder of the view to drag.
         */
        fun onStartDrag(viewHolder: RecyclerView.ViewHolder?)
    }

    companion object {
        const val TAG = "RecyclerViewFragment"
    }
}